crates/datasources/src/lib.rs

```diff
@@ -17,3 +30,49 @@ pub mod postgres;
 pub mod snowflake;
 pub mod sqlite;
 pub mod sqlserver;
+
+
+pub type DatasourceError = Box<dyn std::error::Error + Send + Sync>;
+#[async_trait]
+pub trait Datasource: Send + Sync {
+    fn name(&self) -> &'static str;
+
+    /// Validate the provided tunnel options.
+    fn validate_tunnel_connections(
+        &self,
+        tunnel_opts: Option<&TunnelOptions>,
+    ) -> Result<(), DatasourceError>;
+
+    /// Validate the provided credentials.
+    fn validate_credentials(
+        &self,
+        creds: Option<CredentialsOptions>,
+    ) -> Result<(), DatasourceError>;
+
+    /// Create a new datasource from the provided table options and credentials.
+    /// CREATE EXTERNAL TABLE foo FROM <name> OPTIONS (...) [CREDENTIALS] (...) [TUNNEL] (...)
+    // TODO: the datasource should have control over it's own CredentialsOptions and TunnelOptions
+    fn table_options_from_stmt(
+        &self,
+        opts: &mut StatementOptions,
+        creds: Option<CredentialsOptions>,
+        tunnel_opts: Option<TunnelOptions>,
+    ) -> Result<TableOptions, DatasourceError>;
+
+
+    async fn create_table_provider(
+        &self,
+        tbl_options: &TableOptions,
+        _tunnel_opts: Option<&TunnelOptions>,
+    ) -> Result<Arc<dyn TableProvider>, DatasourceError>;
+
+    /// Create a new datasource from the provided database options and credentials.
+    /// If the datasource does not support databases, return `Ok(None)`.
+    async fn table_provider_from_db_options(
```

Comment:
do implementations need to create two factories for tables for both db configs (which many data sources aren't going to have)?

it seems reasonable that `Database Options` could have some kind of converter to table options in some cases and this could be an implementation detail of the registry/dispatcher.

Reply:
I definitely understand the sentiment, but this refactor also includes the datasource interface, and implies/suggests the design of that system, so I think it's important that we agree about where we're going with that design at this stage.